presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/IndexedPriorityQueue.java

```diff
@@ -146,4 +147,59 @@ public long getGeneration()
             return generation;
         }
     }
+
+    private static class IndexedPriorityQueueIterator<E>
+            implements Iterator<E>
+    {
+        private final Map<E, Entry<E>> index;
+        private final Iterator<E> queueIterator;
+
+        private boolean hasNext;
+        private boolean needsNext = true;
+        private boolean removeCalled = true;
+        private E value;
+
+        public IndexedPriorityQueueIterator(Set<Entry<E>> queue, Map<E, Entry<E>> index)
+        {
+            this.index = requireNonNull(index, "index is null");
+            requireNonNull(queue, "queue is null");
+            this.queueIterator = transform(queue.iterator(), Entry::getValue);
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            if (!needsNext) {
+                return hasNext;
+            }
+            hasNext = queueIterator.hasNext();
+            if (hasNext) {
+                value = queueIterator.next();
+            }
+            needsNext = false;
+            return hasNext;
+        }
+
+        @Override
+        public E next()
+        {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            removeCalled = false;
+            needsNext = true;
+            return value;
+        }
+
+        @Override
+        public void remove()
+        {
+            if (removeCalled) {
+                throw new IllegalStateException();
+            }
+            removeCalled = true;
+            index.remove(value);
```

Comment:
As implemented in `poll` above - 

```
checkState(index.remove(value) != null, "Failed to remove entry from index");
```

this will help ensure that the state of this class is maintained as expected